File: assets/codebase/menu/NintendoDsReturnOverlayComponent.cs
namespace city.menu {
    /// <summary>
    /// Owns Nintendo DS companion-scene return behavior from the bottom-screen back overlay.
    /// </summary>
    public sealed class NintendoDsReturnOverlayComponent : UpdateComponent {
        /// <summary>
        /// Stable logical scene id used by the demo-disc main menu.
        /// </summary>
        public const string MainMenuSceneId = "DemoDiscMainMenu";

        /// <summary>
        /// Interactable host used by the bottom-screen back button.
        /// </summary>
        InteractableComponent BoundInteractable;

        /// <summary>
        /// Tracks whether the active pointer press began inside the bound interactable.
        /// </summary>
        bool PointerPressStartedInside;

        /// <summary>
        /// Tracks whether this component already requested the return transition during its current lifetime.
        /// </summary>
        bool SceneLoadWasRequested;

        /// <summary>
        /// Binds the sibling interactable when the component is attached to the back-button host.
        /// </summary>
        /// <param name="entity">Owning entity.</param>
        public override void ComponentAdded(Entity entity) {
            base.ComponentAdded(entity);
            TryBindInteractable();
        }

        /// <summary>
        /// Performs per-frame polling for the Nintendo DS back bind.
        /// </summary>
        public override void Update() {
            TryBindInteractable();
            InputSystem inputSystem = Core.Instance.Input;
            if (WasGamepadReturnPressed(inputSystem)) {
                LoadResolvedMainMenuScene();
            }
        }

        /// <summary>
        /// Releases the sibling interactable subscription before the component instance is deleted.
        /// </summary>
        public void Dispose() {
            UnbindInteractable();
        }

        /// <summary>
        /// Releases the sibling interactable subscription when the component detaches from its owner.
        /// </summary>
        /// <param name="entity">Owning entity.</param>
        public override void ComponentRemoved(Entity entity) {
            UnbindInteractable();
            base.ComponentRemoved(entity);
        }

        /// <summary>
        /// Binds the sibling interactable used to receive pointer clicks from the back button host.
        /// </summary>
        void TryBindInteractable() {
            if (BoundInteractable != null) {
                return;
            } else if (Parent == null || Parent.Components == null) {
                return;
            }

            for (int componentIndex = 0; componentIndex < Parent.Components.Count; componentIndex++) {
                if (Parent.Components[componentIndex] is InteractableComponent interactable) {
                    BoundInteractable = interactable;
                    BoundInteractable.CursorEvent += HandleCursorEvent;
                    return;
                }
            }

            throw new InvalidOperationException("NintendoDsReturnOverlayComponent requires a sibling InteractableComponent.");
        }

        /// <summary>
        /// Releases the current interactable binding and clears active press state.
        /// </summary>
        void UnbindInteractable() {
            if (BoundInteractable == null) {
                return;
            }

            BoundInteractable.CursorEvent -= HandleCursorEvent;
            BoundInteractable = null;
            PointerPressStartedInside = false;
        }

        /// <summary>
        /// Handles pointer press and release events from the back button interactable.
        /// </summary>
        /// <param name="relativePosition">Pointer position relative to the interactable.</param>
        /// <param name="delta">Pointer delta reported by the shared interaction router.</param>
        /// <param name="interaction">Current pointer interaction state.</param>
        void HandleCursorEvent(int2 relativePosition, int2 delta, PointerInteraction interaction) {
            if (interaction == PointerInteraction.Press) {
                PointerPressStartedInside = true;
                return;
            }
            if (interaction == PointerInteraction.Release) {
                bool shouldReturnToMenu = PointerPressStartedInside;
                PointerPressStartedInside = false;
                if (shouldReturnToMenu) {
                    LoadResolvedMainMenuScene();
                }
                return;
            }
            if (interaction == PointerInteraction.Leave) {
                PointerPressStartedInside = false;
            }
        }

        /// <summary>
        /// Returns whether the current frame pressed the configured standard platform return action.
        /// </summary>
        /// <param name="inputSystem">Input system supplying the current frame state.</param>
        /// <returns>True when the configured standard platform return action was pressed this frame.</returns>
        bool WasGamepadReturnPressed(InputSystem inputSystem) {
            if (inputSystem == null) {
                throw new ArgumentNullException(nameof(inputSystem));
            }
            if (Core.Instance == null) {
                throw new InvalidOperationException("A core instance must exist before querying the standard platform return action.");
            }

            return Core.Instance.StandardPlatformInput.WasActionPressed(StandardPlatformAction.Return);
        }

        /// <summary>
        /// Loads the logical main menu scene after resolving any active scene-map override.
        /// </summary>
        void LoadResolvedMainMenuScene() {
            if (SceneLoadWasRequested) {
                return;
            }
            if (Core.Instance == null) {
                throw new InvalidOperationException("A core instance must exist before returning to the main menu.");
            }
            if (Core.Instance.SceneManager == null) {
                throw new InvalidOperationException("Core scene manager must be initialized before runtime menu scene loading can occur.");
            }

            string resolvedSceneId = SceneMapComponent.ResolveSceneId(MainMenuSceneId);
            SceneLoadWasRequested = true;
            Core.Instance.SceneManager.LoadScene(resolvedSceneId, SceneLoadMode.Single);
        }
    }
}

File: windows-build-20260602-cube-test-ground-high-static-camera/cooked/code/gameplay/NintendoDsReturnOverlayComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "NintendoDsReturnOverlayComponent.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_string.hpp"
#include "NintendoDsReturnOverlayComponent.hpp"
#include "runtime/native_list.hpp"
#include "system/action.hpp"
#include "Component.hpp"
#include "Core.hpp"
#include "Entity.hpp"
#include "InteractableComponent.hpp"
#include "PointerInteraction.hpp"
#include "SceneLoadMode.hpp"
#include "SceneManager.hpp"
#include "SceneMapComponent.hpp"
#include "StandardPlatformAction.hpp"
#include "StandardPlatformInput.hpp"
#include "UpdateComponent.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"

NintendoDsReturnOverlayComponent::NintendoDsReturnOverlayComponent() : BoundInteractable(), PointerPressStartedInside(), SceneLoadWasRequested()
{
}

void NintendoDsReturnOverlayComponent::ComponentAdded(Entity* entity)
{
UpdateComponent::ComponentAdded(entity);
this->TryBindInteractable();
}

void NintendoDsReturnOverlayComponent::ComponentRemoved(Entity* entity)
{
this->UnbindInteractable();
UpdateComponent::ComponentRemoved(entity);
}

void NintendoDsReturnOverlayComponent::Dispose()
{
this->UnbindInteractable();
}

void NintendoDsReturnOverlayComponent::Update()
{
this->TryBindInteractable();
InputSystem *inputSystem = Core::get_Instance()->get_Input();
    if (this->WasGamepadReturnPressed(inputSystem))
    {
this->LoadResolvedMainMenuScene();
    }
}

void NintendoDsReturnOverlayComponent::HandleCursorEvent(int2 relativePosition, int2 delta, PointerInteraction interaction)
{
    if (interaction == PointerInteraction::Press)
    {
this->PointerPressStartedInside = true;
return;    }
    if (interaction == PointerInteraction::Release)
    {
const bool shouldReturnToMenu = this->PointerPressStartedInside;
this->PointerPressStartedInside = false;
    if (shouldReturnToMenu)
    {
this->LoadResolvedMainMenuScene();
    }
return;    }
    if (interaction == PointerInteraction::Leave)
    {
this->PointerPressStartedInside = false;
    }
}

void NintendoDsReturnOverlayComponent::LoadResolvedMainMenuScene()
{
    if (this->SceneLoadWasRequested)
    {
return;    }
    if (Core::get_Instance() == nullptr)
    {
throw new InvalidOperationException("A core instance must exist before returning to the main menu.");
    }
    if (Core::get_Instance()->get_SceneManager() == nullptr)
    {
throw new InvalidOperationException("Core scene manager must be initialized before runtime menu scene loading can occur.");
    }
const std::string resolvedSceneId = SceneMapComponent::ResolveSceneId(MainMenuSceneId);
this->SceneLoadWasRequested = true;
Core::get_Instance()->get_SceneManager()->LoadScene(resolvedSceneId, SceneLoadMode::Single);
}

void NintendoDsReturnOverlayComponent::TryBindInteractable()
{
    if (this->BoundInteractable != nullptr)
    {
return;    }
else {
    if (this->get_Parent() == nullptr || this->get_Parent()->get_Components() == nullptr)
    {
return;    }
}
for (int32_t componentIndex = 0; componentIndex < this->get_Parent()->get_Components()->get_Count(); componentIndex++) {
    InteractableComponent* interactable = he_cpp_try_cast<InteractableComponent>((*this->get_Parent()->get_Components()).get_Item(componentIndex));
    if (interactable != nullptr)
    {
this->BoundInteractable = interactable;
this->BoundInteractable->CursorEvent += &NintendoDsReturnOverlayComponent::HandleCursorEvent;
return;    }
}
throw new InvalidOperationException("NintendoDsReturnOverlayComponent requires a sibling InteractableComponent.");
}

void NintendoDsReturnOverlayComponent::UnbindInteractable()
{
    if (this->BoundInteractable == nullptr)
    {
return;    }
this->BoundInteractable->CursorEvent -= &NintendoDsReturnOverlayComponent::HandleCursorEvent;
this->BoundInteractable = nullptr;
this->PointerPressStartedInside = false;
}

bool NintendoDsReturnOverlayComponent::WasGamepadReturnPressed(InputSystem* inputSystem)
{
    if (inputSystem == nullptr)
    {
throw new ArgumentNullException("inputSystem");
    }
    if (Core::get_Instance() == nullptr)
    {
throw new InvalidOperationException("A core instance must exist before querying the standard platform return action.");
    }
return Core::get_Instance()->get_StandardPlatformInput()->WasActionPressed(StandardPlatformAction::Return);}

